generic: mtd: add Foresee F35SQA001G/F35SQA002G support

chbgdn opened the pull request:
New Xiaomi AX3000T boards now use Foresee F35SQA001G SPI NAND.
F35SQA002G is supported in upstream, so we can just backport it and add pending patch to support F35SQA001G.

Sent upstream: https://patchwork.ozlabs.org/project/linux-mtd/patch/20241110205048.1751377-1-chbgdn@gmail.com/

chbgdn committed:
generic: mtd: spinand: add support for FORESEE F35SQA002G

Add support for FORESEE F35SQA002G SPI NAND.
Datasheet:
  https://www.longsys.com/uploads/LM-00006FORESEEF35SQA002GDatasheet_1650183701.pdf

Refresh all affected patch with make target/linux/refresh.

Signed-off-by: Martin Kurbanov <mmkurbanov@salutedevices.com>
Signed-off-by: Miquel Raynal <miquel.raynal@bootlin.com>
Signed-off-by: Bohdan Chubuk <chbgdn@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/16915
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

onlyfly34 commented:
I remember that in F35SQA001G, 32H command has to be issued before 34H according to datasheet:
![image](https://github.com/user-attachments/assets/b6f7f781-e89d-49e4-93ff-bf06e68b4f82)

This may cause problems for short data transfer with such SPI-NAND.

chbgdn commented:
> I remember that in F35SQA001G, 32H command has to be issued before 34H according to [datasheet](https://cdn.ozdisan.com/ETicaret_Dosya/704795_871495.pdf): ![image](https://private-user-images.githubusercontent.com/13806417/384870921-b6f7f781-e89d-49e4-93ff-bf06e68b4f82.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEzMzExNDIsIm5iZiI6MTczMTMzMDg0MiwicGF0aCI6Ii8xMzgwNjQxNy8zODQ4NzA5MjEtYjZmN2Y3ODEtZTg5ZC00OWU0LTkzZmYtYmYwNmU2OGI0ZjgyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTExVDEzMTQwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZjNGFhNzkxMTlmMmVjNmY4ZmU3NjJkNjQ4ZGRmMDY2NDM3NmFhOTc1NTFlOTVmNzYxYmViY2JlZmY2OTFkYzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.lXYQr6u3KXpnkiEEy9Vc1_3V-l7cFMouZO7V47q8fe0)
> 
> This may cause problems for short data transfer with such SPI-NAND.

So we need this patch?
```
diff --git a/drivers/mtd/nand/spi/foresee.c b/drivers/mtd/nand/spi/foresee.c
index e0d2d9257..05b27341b 100644
--- a/drivers/mtd/nand/spi/foresee.c
+++ b/drivers/mtd/nand/spi/foresee.c
@@ -22,8 +22,8 @@ static SPINAND_OP_VARIANTS(write_cache_variants,
 		SPINAND_PROG_LOAD(true, 0, NULL, 0));
 
 static SPINAND_OP_VARIANTS(update_cache_variants,
-		SPINAND_PROG_LOAD_X4(false, 0, NULL, 0),
-		SPINAND_PROG_LOAD(false, 0, NULL, 0));
+		SPINAND_PROG_LOAD_X4(true, 0, NULL, 0),
+		SPINAND_PROG_LOAD(true, 0, NULL, 0));
 
 static int f35sqa002g_ooblayout_ecc(struct mtd_info *mtd, int section,
 				    struct mtd_oob_region *region)
```

onlyfly34 commented:
@chbgdn
I saw that on newest mainline. However, current `foresee.c` will always use 32H for `update_cache` operations, which means it's not compitable with those SPI controller who can't transmit 2048 bytes (most small-density SPI-NAND's page size nowadays) at one time. 

The following controller's driver seems that they can't transmit 2048 bytes in one transmission:
- spi-amd.c: 64 bytes [AMD_SPI_MAX_DATA](https://elixir.bootlin.com/linux/v6.11.7/C/ident/AMD_SPI_MAX_DATA)
- spi-amlogic-spifc-a1.c: 512 bytes [SPIFC_A1_BUFFER_SIZE](https://elixir.bootlin.com/linux/v6.11.7/C/ident/SPIFC_A1_BUFFER_SIZE)
- spi-fsl-qspi.c: 1KB
- spi-hisi-sfc-v3xx.c: 64*6 bytes
- spi-intel.c: 64 bytes [INTEL_SPI_FIFO_SZ](https://elixir.bootlin.com/linux/v6.11.7/C/ident/INTEL_SPI_FIFO_SZ)
- spi-microchip-core-qspi.c: 256 bytesc [MAX_DATA_CMD_LEN](https://elixir.bootlin.com/linux/v6.11.7/C/ident/MAX_DATA_CMD_LEN)
- spi-nxp-fspi.c: TX:1KB, RX: 512B in FIFO mode
- spi-wpcm-fiu.c: 4B

I guess we need to add some check in Linux upstream code to make F35SQA001G compatible in spi-mem framework.

chbgdn commented:
@onlyfly34 Thanks for explanation. I think we can left all as is in this PR, merge it and after solving spi-mem framework related issues in upstream backport it too.

I'm using newest revision of AX3000T (F35SQA001G SPI NAND, AN8855 switch) flashed OpenWRT with patches from this PR and #16709, all working flawless for few days.

nstepanets commented:
> I remember that in F35SQA001G, 32H command has to be issued before 34H according to [datasheet](https://cdn.ozdisan.com/ETicaret_Dosya/704795_871495.pdf)

It looks like the description in the datasheet is ambiguous. If every 34H command has to be preceded by a 32H command, then there is no point in its existence...

Ansuel commented:
@chbgdn patch looks good... I would stretch this till the device is actually supported + support for an8855. If we are lucky we have the patch accepted upstream and we can better handle this.

chbgdn commented:
@Ansuel sure, for now I will left it as is. If my patch will be upstreamed soon - we can handle this just with one backport from newer kernel version.

onlyfly34 commented:
@chbgdn @Ansuel

About this issue, I submitted a patch to Linux upstream:
https://patchwork.ozlabs.org/project/linux-mtd/patch/20241119093949.3014-1-SkyLake.Huang@mediatek.com/

onlyfly34 commented:
> > I remember that in F35SQA001G, 32H command has to be issued before 34H according to [datasheet](https://cdn.ozdisan.com/ETicaret_Dosya/704795_871495.pdf)
> 
> It looks like the description in the datasheet is ambiguous. If every 34H command has to be preceded by a 32H command, then there is no point in its existence...

Not every 34H. It's 32H+34H+34H+34H+34H.... First 32H is necessary.

Ansuel commented:
@chbgdn thanks a lot for the efforts!

hauke commented:
How do we handle the patch from @onlyfly34 ?
https://patchwork.ozlabs.org/project/linux-mtd/patch/20241119093949.3014-1-SkyLake.Huang@mediatek.com/
It got some comment from the mtd maintainer.

I would merge this PR even without this fix as it is also not in mainline Linux yet. 

I am concerned that some vendors will use this flash chip when it is cheap.

chbgdn committed:
generic: mtd: spinand: add support for FORESEE F35SQA001G

Add support for FORESEE F35SQA001G SPI NAND.

Similar to F35SQA002G, but differs in capacity.
Datasheet:
  -  https://cdn.ozdisan.com/ETicaret_Dosya/704795_871495.pdf

Tested on Xiaomi AX3000T flashed with OpenWRT.

Signed-off-by: Bohdan Chubuk <chbgdn@gmail.com>
Signed-off-by: Miquel Raynal <miquel.raynal@bootlin.com>
Link: https://github.com/openwrt/openwrt/pull/16915
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

chbgdn commented:
@Ansuel @hauke Moved second patch to backports and rebased on top of main branch.

nstepanets commented:
> First 32H is necessary.

If I understood correctly, then:
- The current implementation using 84H/34H is correct and should theoretically work with SPI NAND from any available memory manufacturer.
- For Foresee SPI NAND, the command sequence must start with 02H / 32H instead of 84H / 34H (recommendation from the Foresee datasheet), otherwise it may encounter program failure issue.
- If the SPI controller cannot transfer 2048 bytes (size of the SPI NAND internal buffer) at once, then subsequent commands must be 84H or 34H as usual.

remittor commented:
@chbgdn 
Please, add this patch to `package/boot/uboot-mediatek/patches/`
